### PR TITLE
[PLC] [Test] [Evaluation] Failure handling

### PR DESCRIPTION
**Pull request by effectfully**
The builtins machinery allows to handle exceptions (i.e. `Error`) _in_ Plutus Core. Didn't expect that, huh?

Unfortunately, the CEK machine does not support this use case. It can be tweaked, but I didn't find a reasonable way of doing so (see comments to the PR), so I do not propose to merge this PR as is, instead I'll rollback the changes to the machine and update the tests just to have them anyway.

(I've written the `dynamicCatchFailure` test case, because I thought the costing PR might break it, but given that the CEK machine doesn't support this use case, no PR can break what is already broken)

It's quite an annoying edge case that the builtins machinery allows to express a certain thing, but the evaluator doesn't support that thing. We can unsupport failulre handling by splitting the `KnownType` type class into two `MakeKnown` and `ReadKnown` classes, but it's ugly and we'll need to delete a couple of existing tests that pass. Not sure what to do here.

### CODE REVIEW

File: language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs [hunk 1]
```diff
@@ -135,16 +135,18 @@ computeCek con (Var _ varName)          = do
...
-    withVarEnv argVarEnv $ computeCek (FrameApplyFun funVarEnv fun : con) arg
-returnCek (FrameApplyFun funVarEnv fun : con) arg = do
+returnCek []                                  Error{}     = pure EvaluationFailure
+returnCek []                                  res         = pure $ EvaluationSuccess res
+returnCek (FrameApplyFun funVarEnv fun : con) arg         = do
     argVarEnv <- getVarEnv
     applyEvaluate funVarEnv argVarEnv con fun arg
-returnCek (FrameIWrap ann pat arg      : con) val = returnCek con $ IWrap ann pat arg val
-returnCek (FrameUnwrap                 : con) dat = case dat of
+returnCek (_ : con)                           err@Error{} = returnCek con err
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs]: So the idea is that we do not throw an `Error` immediately, but give the `FrameApplyFun` frame a chance to handle the error. For any other frame we let the error pop up, but we do not change the type that `Error` carries, so this breaks subjection reduction (don't tell Phil I've written something like that!). If the evaluator was untyped, I believe it would make sense to adapt the CEK machine this way, but breaking subject reduction is a too terrible of a sin.

Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs]: Note that we already don't return terms whose internal types are correct because we don't actually do type substitutions, e.g. https://github.com/input-output-hk/plutus/issues/1145#issuecomment-503954247. So I'm not sure how much of a sin this is.

Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs]: Note that we already don't return terms whose internal types are correct because we don't actually do type substitutions, e.g. https://github.com/input-output-hk/plutus/issues/1145#issuecomment-503954247. So I'm not sure how much of a sin this is.

**Comment by michaelpj** [comment]
> The builtins machinery allows to handle exceptions (i.e. Error) in Plutus Core. Didn't expect that, huh?

Do we *want* this?

File: language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs (hunk 1 above)
Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs]: > Note that we already don't return terms whose internal types are correct because we don't actually do type substitutions

Yeah, but I think it's also a bug. But that bug is easily fixable unlike the one that I introduced here.

Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs]: Hmmm, I'm not sure. See also https://github.com/input-output-hk/plutus/issues/948

**Comment by kwxm** [comment]
> > The builtins machinery allows to handle exceptions (i.e. Error) in Plutus Core. Didn't expect that, huh?
> 
> Do we _want_ this?

We could just tell people not to do it. There's lots of other things that implementers of extensible built-ins shouldn't do (I/O, random number generation, ...). Presumably we'll eventually have a specification of this stuff, and it should say that if you do certain things then you're in danger of breaking everything.

**Comment by effectfully** [comment]
> Do we _want_ this?

Hard to tell. Most likely, no, but you know how people like throwing errors in libraries that eventually pop up to the user when the user doesn't really care about some particular error happening and just wants the try-again-later behavior or the who-cares-that-it-failed one. But then our setting is quite different. But we may want to use `catchFailure` when there is a risk of diving by zero. Hard to tell.

The problem is that it's hard not to support this in the builtins machinery, so I dislike the discrepancy between the builtins machinery that pretends something can be done and the evaluation machinery that doesn't actually allow that.

I think, right now, we don't want this. We may want it in future and it's nice to have a proof that this kind of thing can be done, but probably the best way forward is to rollback the changes to the CEK machine and add a comment "don't think you can actually handle exceptions".

**Comment by michaelpj** [comment]
> Most likely, no, but you know how people like throwing errors in libraries that eventually pop up to the user when the user doesn't really care about some particular error happening and just wants the try-again-later behavior or the who-cares-that-it-failed one. But then our setting is quite different.

Yes, this is about *builtins* not random library code. I really don't think we should let people catch `error` ever, it makes it much more complex.

**Comment by effectfully** [comment]
> Yes, this is about builtins not random library code. 

I mean, someone wrote a library and it throws an `Error` and breaks your code, so you might decide to fix it by using `catchFailure`. We could make it really expensive to make people not use it unless they really need to :)

> I don't think we should let people catch `error` ever, it makes it much more complex.

I generally agree. As I said, I'm in favor of not changing the CEK machine to support this use case.

**Comment by michaelpj** [comment]
> I mean, someone wrote a library and it throws an Error and breaks your code, so you might decide to fix it by using catchFailure. We could make it really expensive to make people not use it unless they really need to :)

If we make it possible, people will do it...

**Comment by effectfully** [comment]
Ready for review.

File: language-plutus-core/test/Evaluation/DynamicBuiltins/Definition.hs [hunk 2]
```diff
@@ -121,10 +123,122 @@ test_dynamicReverse =
         lhs === Right (Right (EvaluationSuccess . PlcList $ Prelude.reverse is))
         lhs === rhs
 
+dynamicCatchErrorName :: DynamicBuiltinName
```
Comment (michaelpj) [inline language-plutus-core/test/Evaluation/DynamicBuiltins/Definition.hs]: :see_no_evil:

**Comment by michaelpj** [comment]
Actually, maybe a note in one of the machines about the way that error interacts with builtins would be a good idea?

**Comment by effectfully** [comment]
#1813 unsupports error handling with builtins, so this PR is mostly pointless. I rolled back the rolled back changes just to have a full working solution here and now this PR can be closed. There is one useful test, though, I'll add it separately.

**Comment by effectfully** [comment]
> There is one useful test, though, I'll add it separately.

Incorporated it into #1813 just not to disseminate redundant PRs.
